src/protocol/boolean/random_bits_generator.rs

```diff
@@ -0,0 +1,139 @@
+use super::solved_bits::{RandomBitsShare, SolvedBits};
+use crate::error::Error;
+use crate::ff::Field;
+use crate::protocol::context::SemiHonestContext;
+use crate::protocol::RecordId;
+use std::collections::HashMap;
+
+/// A struct that pre-generates and buffers random sharings of bits from the
+/// `SolvedBits` protocol. Any protocol who wish to use a random-bits can draw
+/// one by calling `take_one()`. It will call `SolvedBits` once the stock falls
+/// below `REFILL_THRESHOLD` until it fills up the empty slots.
+///
+/// Internally, it uses two pointers. `read_pointer` keeps track of the next
+/// item to return, and `write_pointer` points to the least recent empty slot.
+/// The pointers are `i8`. Incrementing them cause wrap around at the boundary,
+/// which makes the `buffer` hash map a ring buffer.
+#[allow(dead_code)]
+pub struct RandomBitsGenerator<'a, F: Field> {
+    context: SemiHonestContext<'a, F>,
+    buffer: HashMap<i8, RandomBitsShare<F>>,
+    counter: u32,
+    read_pointer: i8,
+    write_pointer: i8,
+    abort_count: u32,
+}
+
+#[allow(dead_code)]
+impl<'a, F: Field> RandomBitsGenerator<'a, F> {
+    const REFILL_THRESHOLD: u8 = 16;
+
+    pub fn new(context: SemiHonestContext<'a, F>) -> Self {
+        Self {
+            context,
+            buffer: HashMap::with_capacity(u8::MAX.into()),
+            counter: 0,
+            read_pointer: 0,
+            write_pointer: 0,
+            abort_count: 0,
+        }
+    }
+
+    /// Takes one `RandomBitsShare` instance out of the buffer. If the number
+    /// of buffered items fall below a threshold, it'll replenish.
+    pub async fn take_one(&mut self) -> Result<RandomBitsShare<F>, Error> {
+        self.refill_if_needed().await?;
+        let r = self
+            .buffer
+            .remove(&self.read_pointer)
+            .expect("unexpected buffer depletion");
+        self.read_pointer = self.read_pointer.wrapping_add(1);
```

Comment:
What does `wrapping_add(1)` do on a `i8`? When does it wrap around? Does 127 wrap around to -128 or something?

Reply:
Yes, that is correct. I've changed everything to `u8`.